qemu/tests/cpu_topology_test.py

```diff
@@ -0,0 +1,75 @@
+import re
+import random
+
+from avocado.utils import cpu
+
+from virttest import error_context
+from virttest import env_process
+
+from provider.cpu_utils import check_guest_cpu_topology
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Check guest gets correct vcpu num, cpu cores, processors, sockets, siblings
+
+    1) Boot guest with options: -smp n,cores=x,threads=y,sockets=z...
+    2) Check cpu topology
+
+    :param test: QEMU test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+
+    def check(p_name, exp, check_cmd):
+        """
+        Check the cpu property inside guest
+        :param p_name: Property name
+        :param exp: The expect value
+        :param check_cmd: The param to get check command
+        """
+        res = session.cmd_output(params[check_cmd]).strip()
+        if int(res) != int(exp):
+            test.fail('The vcpu %s number inside guest is %s,'
+                      ' while it is set to %s' % (p_name, res, exp))
+
+    vm_name = params['main_vm']
+    os_type = params['os_type']
+    vcpu_threads_list = [1, 2]
+    if params['machine_type'] == 'pseries':
+        vcpu_threads_list = [1, 2, 4, 8]
+    params['vcpu_cores'] = vcpu_cores = random.randint(1, 10)
+    host_cpu = cpu.online_count()
+    for vcpu_threads in vcpu_threads_list:
+        vcpu_sockets = min(max(host_cpu * 5 // (vcpu_cores * vcpu_threads), 1),
+                           random.randint(1, 10))
+        vcpu_sockets = 2 if (os_type == 'Windows' and
+                             vcpu_sockets > 2) else vcpu_sockets
+        params['vcpu_sockets'] = vcpu_sockets
+        params['vcpu_threads'] = vcpu_threads
+        params['smp'] = params['vcpu_maxcpus'] = (vcpu_cores *
+                                                  vcpu_threads * vcpu_sockets)
+        params['start_vm'] = 'yes'
+        try:
+            env_process.preprocess_vm(test, params, env, vm_name)
+        except Exception as e:
+            # The cpu topology sometimes will be changed by
+            # qemu_vm.VM.make_create_command, and thus cause qemu vm fail to
+            # start, which is expected; Modify the value and restart vm in
+            # this case, and verify cpu topology inside guest after that
```

Comment:
Does this mean we won't fix the auto AMD bug like yihuang said "IIUC, AMD CPUs do not support multiple threads except EYPC" ?

Reply:
@nanliu-r No, I will fix the EYPC issue, but it should be independent of this case, since you mention that intel has issues sometimes, so here it can handle all exceptions I think, including those we don't meet yet.

Reply:
@nanliu-r Fixed it here: https://github.com/avocado-framework/avocado-vt/pull/2717